File: organized_code/STM Paper 2017/Matlab/Data Visualization/Fitted_Poles.m
clear all
close all
clc

%format long e

%% Normals

MCADataValues = xlsread('../../Data/Processed/CAT_Normals','Fits','A2:M21');

Normalnum = MCADataValues(:,1);
k0 = MCADataValues(:,2);
k1 = MCADataValues(:,3);
k2 = MCADataValues(:,4);
kn = MCADataValues(:,5);
kd = MCADataValues(:,6);

II_0 = MCADataValues(:,7);
V_0 = MCADataValues(:,8);
VII_0 = MCADataValues(:,9);
VIII_0 = MCADataValues(:,10);
IX_0 = MCADataValues(:,11);
X_0 = MCADataValues(:,12);
ATIII_0 = MCADataValues(:,13);

s = tf('s');



for j=1:length(k0)
    sysnodel(j) = kn(j)/(s^3 + k2(j)*s^2 + k1(j)*s + k0(j));
    sysdel(j) = sysnodel(j)*exp(-kd(j)*s);
    [r(:,j) p(:,j) direct] = residue(kn(j),[1 k2(j) k1(j) k0(j)]);
    for i = 1:3
        checkpoleisonlyreal(i,j) = isreal(p(i,j));
        if isreal(p(i,j))
            onlyrealpoles(i,j) = p(i,j);
        else
            onlypoleswithcomplexparts(i,j) = p(i,j);
            realpartofpoleswithcomplexparts(i,j) = real(p(i,j));
            complexpartofpoleswithcomplexparts(i,j)= imag(p(i,j));
        end
    end
end



numonlyrealpoles = sum(checkpoleisonlyreal);


for j=1:length(k0)
    if numonlyrealpoles(j)==1
        if min(realpartofpoleswithcomplexparts(:,j)) < min(onlyrealpoles(:,j))
            checkpoleisrightofpair(j) = 1;
        else
            checkpoleisrightofpair(j) = 0;
        end
    else
        checkpoleisrightofpair(j) = NaN;
    end
end


for j=1:length(k0)
    if checkpoleisrightofpair(j) == 1
        rightrealpolevalue(j) = min(onlyrealpoles(:,j));
        leftpairrealpolevalue(j) = min(realpartofpoleswithcomplexparts(:,j));
        leftpairimagpolevalue(j) = min(complexpartofpoleswithcomplexparts(:,j));
        leftpairmagpolevalue(j) = sqrt(leftpairrealpolevalue(j)^2 + leftpairimagpolevalue(j)^2);
        leftpairanglepolevalue(j) = atan2(leftpairimagpolevalue(j),leftpairrealpolevalue(j));
    else
        rightrealpolevalue(j) = NaN;
        leftpairrealpolevalue(j) = NaN;
        leftpairimagpolevalue(j) = NaN;
        leftpairmagpolevalue(j) = NaN;
        leftpairanglepolevalue(j) = NaN;
    end
end
    

figure(1)
clf
subplot(2,2,1)
hold on

box on


plotmkcolors = {'r' 'g' 'b' 'k' 'm' 'c' 'y'};
plotmkstyles = {'^' 's' 'd' 'x' 'o' '*' '+' 'v' '<' '>' 'p' 'h'};

mkcolorctr = 1;
mkstylectr = 1;


for j = 1:length(k0)
    plot(real(p(:,j)),imag(p(:,j)),[char(plotmkcolors(mkcolorctr)),char(plotmkstyles(mkstylectr))],'LineWidth',6,'MarkerSize',12);
    mkcolorctr = mkcolorctr + 1;
    if mkcolorctr > length(plotmkcolors)
        mkcolorctr = 1;
        mkstylectr = mkstylectr + 1;
    end    
end
xlabel('Real Axis')
ylabel('Imaginary Axis')
%title('All Pole Locations for All Normal Patients')
axis([-0.6 0.043 -0.857 0.857]);
% plot([-0.75,0.05],[0,0],'k:');
% plot([0,0],[-0.75,0.75],'k:');
% axis([-1.4 0.1 -2 2]); %same as trauma plot
plot([-1.4,0.1],[0,0],'k:');
plot([0,0],[-2,2],'k:');
%title('A')
figureHandle = gcf;
%# make all text in the figure to size 30 and bold
set(findall(figureHandle,'type','text'),'fontSize',27)%,'fontWeight','bold')
set(gca,'FontSize',27)


AverageRightRealPoleValue = mean(rightrealpolevalue)
AverageLeftPairMagPoleValue = mean(leftpairmagpolevalue)
AverageLeftPairAnglePoleValue = mean(leftpairanglepolevalue)
AverageDelay = mean(kd)

%% Trauma Patients

%MCADataValues = xlsread('../../Data/Processed/CAT_Trauma','Fits','B2:M48');
MCADataValues = xlsread('../../Data/Processed/CAT_Trauma','Fits','B2:M41');

%Normalnum = MCADataValues(:,1);
k0 = MCADataValues(:,1);
k1 = MCADataValues(:,2);
k2 = MCADataValues(:,3);
kn = MCADataValues(:,4);
kd = MCADataValues(:,5);

% II_0 = MCADataValues(:,7);
% V_0 = MCADataValues(:,8);
% VII_0 = MCADataValues(:,9);
% VIII_0 = MCADataValues(:,10);
% IX_0 = MCADataValues(:,11);
% X_0 = MCADataValues(:,12);
% ATIII_0 = MCADataValues(:,13);

s = tf('s');

%%

for j=1:length(k0)
    sysnodel(j) = kn(j)/(s^3 + k2(j)*s^2 + k1(j)*s + k0(j));
    sysdel(j) = sysnodel(j)*exp(-kd(j)*s);
    [r(:,j) p(:,j) direct] = residue(kn(j),[1 k2(j) k1(j) k0(j)]);
    for i = 1:3
        checkpoleisonlyreal(i,j) = isreal(p(i,j));
        if isreal(p(i,j))
            onlyrealpoles(i,j) = p(i,j);
        else
            onlypoleswithcomplexparts(i,j) = p(i,j);
            realpartofpoleswithcomplexparts(i,j) = real(p(i,j));
            complexpartofpoleswithcomplexparts(i,j)= imag(p(i,j));
        end
    end
end



numonlyrealpoles = sum(checkpoleisonlyreal);


for j=1:length(k0)
    if numonlyrealpoles(j)==1
        if min(realpartofpoleswithcomplexparts(:,j)) < min(onlyrealpoles(:,j))
            checkpoleisrightofpair(j) = 1;
        else
            checkpoleisrightofpair(j) = 0;
        end
    else
        checkpoleisrightofpair(j) = NaN;
    end
end



for j=1:length(k0)
    if checkpoleisrightofpair(j) == 1
        rightrealpolevalue(j) = min(onlyrealpoles(:,j));
        leftpairrealpolevalue(j) = min(realpartofpoleswithcomplexparts(:,j));
        leftpairimagpolevalue(j) = min(complexpartofpoleswithcomplexparts(:,j));
        leftpairmagpolevalue(j) = sqrt(leftpairrealpolevalue(j)^2 + leftpairimagpolevalue(j)^2);
        leftpairanglepolevalue(j) = atan2(leftpairimagpolevalue(j),leftpairrealpolevalue(j));
    else
        rightrealpolevalue(j) = NaN;
        leftpairrealpolevalue(j) = NaN;
        leftpairimagpolevalue(j) = NaN;
        leftpairmagpolevalue(j) = NaN;
        leftpairanglepolevalue(j) = NaN;
    end
end
    

%figure(2)
%clf
subplot(2,2,2)
hold on

box on


plotmkcolors = {'r' 'g' 'b' 'k' 'm' 'c' 'y'};
plotmkstyles = {'^' 's' 'd' 'x' 'o' '*' '+' 'v' '<' '>' 'p' 'h'};

mkcolorctr = 1;
mkstylectr = 1;


for j = 1:length(k0)
    plot(real(p(:,j)),imag(p(:,j)),[char(plotmkcolors(mkcolorctr)),char(plotmkstyles(mkstylectr))],'LineWidth',6,'MarkerSize',12);
    mkcolorctr = mkcolorctr + 1;
    if mkcolorctr > length(plotmkcolors)
        mkcolorctr = 1;
        mkstylectr = mkstylectr + 1;
    end    
end
xlabel('Real Axis')
ylabel('Imaginary Axis')
%title('All Pole Locations for All Trauma Patients')
axis([-1.4 0.1 -2 2]);
plot([-1.4,0.1],[0,0],'k:');
plot([0,0],[-2,2],'k:');
%title('B')
figureHandle = gcf;
%# make all text in the figure to size 30 and bold
set(findall(figureHandle,'type','text'),'fontSize',27)%,'fontWeight','bold')
set(gca,'FontSize',27)



%load only data for which there are complete factor values
AvailableData = 1:40;
k0 = MCADataValues(AvailableData,1);
k1 = MCADataValues(AvailableData,2);
k2 = MCADataValues(AvailableData,3);
kn = MCADataValues(AvailableData,4);
kd = MCADataValues(AvailableData,5);

II_0 = MCADataValues(AvailableData,6);
V_0 = MCADataValues(AvailableData,7);
VII_0 = MCADataValues(AvailableData,8);
VIII_0 = MCADataValues(AvailableData,9);
IX_0 = MCADataValues(AvailableData,10);
X_0 = MCADataValues(AvailableData,11);
ATIII_0 = MCADataValues(AvailableData,12);

rightrealpolevalue = rightrealpolevalue(1,AvailableData);
leftpairanglepolevalue = leftpairanglepolevalue(1,AvailableData);
leftpairmagpolevalue = leftpairmagpolevalue(1,AvailableData);

File: organized_code/STM Paper 2017/Python/Data Visualization/Fitted_Poles.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from control.matlab import *
import scipy as spy

# Normals
CATN = '../../Data/Processed/CAT_Normals.xlsx'
dfN = pd.read_excel(CATN, sheet_name='Fits', usecols="A:M").dropna()
MCADataValues = dfN[:20].to_numpy()
# print(MCADataValues)
Normalnum = MCADataValues[:,0]
k0 = MCADataValues[:,1]
k1 = MCADataValues[:,2]
k2 = MCADataValues[:,3]
kn = MCADataValues[:,4]
kd = MCADataValues[:,5]

II_0 = MCADataValues[:,6]
V_0 = MCADataValues[:,7]
VII_0 = MCADataValues[:,8]
VIII_0 = MCADataValues[:,9]
IX_0 = MCADataValues[:,10]
X_0 = MCADataValues[:,11]
ATIII_0 = MCADataValues[:,12]
# print(k0.shape[0])

s = tf('s');
sysnodel = []   # System's transfer functions without delay

r = np.empty((3, k0.shape[0]), dtype=np.cdouble)
p = np.empty((3, k0.shape[0]), dtype=np.cdouble)
onlyrealpoles = np.empty((3, k0.shape[0]))
onlypoleswithcomplexparts = np.empty((3, k0.shape[0]), dtype=np.cdouble)
realpartofpoleswithcomplexparts = np.empty((3, k0.shape[0]))
complexpartofpoleswithcomplexparts = np.empty((3, k0.shape[0]))
checkpoleisonlyreal = np.empty((3, k0.shape[0]), dtype=int)

# Separate real poles and complex poles
# and find complex poles' real and imaginary parts
for j in range(k0.shape[0]):
    sysnodel.append(kn[j]/(s**3 + k2[j] * s**2 + k1[j]*s + k0[j]))
    r[:,j], p[:,j], direct = spy.signal.residue(kn[j],[1, k2[j], k1[j], k0[j]])
    for i in range(3):
        checkpoleisonlyreal[i, j] = np.isreal(p[i,j])
        if np.isreal(p[i,j]):
            onlyrealpoles[i, j] = np.real(p[i, j])
            onlypoleswithcomplexparts[i, j] = complex(np.nan, np.NAN)
            realpartofpoleswithcomplexparts[i, j] = np.nan
            complexpartofpoleswithcomplexparts[i, j] = np.nan
        else:
            onlyrealpoles[i, j] = np.NAN
            onlypoleswithcomplexparts[i, j] = p[i, j]
            realpartofpoleswithcomplexparts[i, j] = np.real(
                onlypoleswithcomplexparts[i, j])
            complexpartofpoleswithcomplexparts[i, j] = np.imag(
                onlypoleswithcomplexparts[i, j])

numonlyrealpoles = sum(checkpoleisonlyreal)
# print(complexpartofpoleswithcomplexparts)
checkpoleisrightofpair = np.empty(k0.shape[0], dtype=int)

for j in range(k0.shape[0]):
    if numonlyrealpoles[j] == 1:
        if np.min(realpartofpoleswithcomplexparts[:,j],
                     where=~np.isnan(realpartofpoleswithcomplexparts[:,j]),
                     initial=1000) < np.min(onlyrealpoles[:,j],
                                            where=~np.isnan(onlyrealpoles[:,j]),
                                            initial=1000):
            # This means the complex pole is more "negative" than the real pole
            checkpoleisrightofpair[j] = 1
        else:
            # This means the complex pole is not on the right side of the real pole
            checkpoleisrightofpair[j] = 0
    else:
        # This means the system only has one real pole and no complex poles
        checkpoleisrightofpair[j] = np.nan


# print(onlyrealpoles)
rightrealpolevalue = np.empty(k0.shape[0])
leftpairrealpolevalue = np.empty(k0.shape[0])
leftpairimagpolevalue = np.empty(k0.shape[0])
leftpairmagpolevalue = np.empty(k0.shape[0])
leftpairanglepolevalue = np.empty(k0.shape[0])

for j in range(k0.shape[0]):
    if checkpoleisrightofpair[j] == 1:
        rightrealpolevalue[j] = np.min(onlyrealpoles[:,j],
                                       where=~np.isnan(onlyrealpoles[:,j]),
                                       initial=1000)
        leftpairrealpolevalue[j] = np.min(realpartofpoleswithcomplexparts[:,j],
                                          where=~np.isnan(realpartofpoleswithcomplexparts[:,j]),
                                          initial=1000)
        leftpairimagpolevalue[j] = np.min(complexpartofpoleswithcomplexparts[:,j],
                                          where=~np.isnan(complexpartofpoleswithcomplexparts[:,j]),
                                          initial=1000)
        leftpairmagpolevalue[j] = np.sqrt([leftpairrealpolevalue[j]**2+
                                           leftpairimagpolevalue[j]**2])[0]
        leftpairanglepolevalue[j] = np.arctan2(leftpairimagpolevalue[j],
                                               leftpairrealpolevalue[j])
    else:
        rightrealpolevalue[j] = np.nan
        leftpairrealpolevalue[j] = np.nan
        leftpairimagpolevalue[j] = np.nan
        leftpairmagpolevalue[j] = np.nan
        leftpairanglepolevalue[j] = np.nan

fh = plt.figure(figsize=(1,2))
plt.subplot(121)
plotmkcolors = ['r','g','b','k','m','c','y']
plotmkstyles = ['^','s','d','x','o','*','+','v','<','>','p','h']
mkcolorctr = 0
mkstylectr = 0
for j in range(k0.shape[0]):
    plt.scatter(np.real(p[:,j]), np.imag(p[:,j]), s=12, c=plotmkcolors[mkcolorctr],
                marker=plotmkstyles[mkstylectr], linewidth=6)
    mkcolorctr += 1
    if mkcolorctr >= len(plotmkcolors):
        mkcolorctr = 0
        mkstylectr = mkstylectr + 0
# print(sysnodel[0])
plt.xlabel('Real Axis')
plt.ylabel('Imaginary Axis')
plt.xlim(-0.6, 0.043)
plt.ylim(-0.857, 0.857)
plt.plot([-1.4,0.1],[0,0],'k:');
plt.plot([0,0],[-2,2],'k:');


plt.show()
